user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserOperator.java

```diff
@@ -139,12 +149,111 @@ protected void createOrUpdate(Reconciliation reconciliation, KafkaUser kafkaUser
             scramAcls = user.getSimpleAclRules();
         }
 
+        KafkaUserStatus userStatus = new KafkaUserStatus();
         CompositeFuture.join(
                 scramShaCredentialOperator.reconcile(user.getName(), password),
-                secretOperations.reconcile(namespace, user.getSecretName(), desired),
+                reconcileSecretAndSetStatus(namespace, user, desired, userStatus),
                 aclOperations.reconcile(KafkaUserModel.getTlsUserName(userName), tlsAcls),
                 aclOperations.reconcile(KafkaUserModel.getScramUserName(userName), scramAcls))
-                .map((Void) null).setHandler(handler);
+                .setHandler(reconciliationResult -> {
+
+                    if (kafkaUser.getMetadata().getGeneration() != null)    {
+                        userStatus.setObservedGeneration(kafkaUser.getMetadata().getGeneration());
+                    }
+
+                    Condition readyCondition = ConditionUtils.buildConditionFromReconciliationResult(reconciliationResult.mapEmpty());
+
+                    userStatus.setUsername(user.getName());
+                    userStatus.setConditions(Collections.singletonList(readyCondition));
+
+                    updateStatus(kafkaUser, reconciliation, userStatus).setHandler(statusResult -> {
+                        if (statusResult.succeeded())    {
+                            log.debug("Status for {} is up to date", kafkaUser.getMetadata().getName());
+                        } else {
+                            log.error("Failed to set status for {}. {}", kafkaUser.getMetadata().getName(), statusResult.cause().getMessage());
+                        }
+
+                        // If both features succeeded, createOrUpdate succeeded as well
+                        // If one or both of them failed, we prefer the reconciliation failure as the main error
+                        if (reconciliationResult.succeeded() && statusResult.succeeded())    {
+                            createOrUpdateFuture.complete();
+                        } else if (reconciliationResult.failed())    {
+                            createOrUpdateFuture.fail(reconciliationResult.cause());
+                        } else {
+                            createOrUpdateFuture.fail(statusResult.cause());
+                        }
+                    });
+                    handler.handle(createOrUpdateFuture);
```

Comment:
Why should this call be here? And not either in the `statusResult` handler or just at the end of the `createOrUpdate` method?